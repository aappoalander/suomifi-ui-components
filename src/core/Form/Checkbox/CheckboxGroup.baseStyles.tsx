import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { element } from '../../theme/reset';

export const baseStyles = (theme: SuomifiTheme) => css`
  ${element(theme)}

  & .fi-checkbox-group_label--visible {
    margin-bottom: 10px;
  }

  & .fi-checkbox-group_legend .fi-hint-text {
    margin-bottom: 10px;
  }

  & .fi-checkbox-group_container {
    & > .fi-checkbox {
      margin-bottom: ${theme.spacing.xs};

      &:last-child {
        margin-bottom: 0;
      }
    }

    & .fi-checkbox--large {
      margin-top: ${theme.spacing.xxs};
      margin-bottom: ${theme.spacing.s};
    }
  }
`;
