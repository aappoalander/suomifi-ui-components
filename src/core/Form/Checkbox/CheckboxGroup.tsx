import React, { Component, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import { SuomifiThemeProp, SuomifiThemeConsumer } from '../../theme';
import { HtmlDiv, HtmlFieldSet, HtmlLegend } from '../../../reset';
import { Label } from '../Label/Label';
import { HintText } from '../HintText/HintText';
import { CheckboxProps } from './Checkbox';
import { baseStyles } from './CheckboxGroup.baseStyles';
import { AutoId } from '../../utils/AutoId/AutoId';
import classnames from 'classnames';

const baseClassName = 'fi-checkbox-group';
const checkboxGroupClassNames = {
  legend: `${baseClassName}_legend`,
  labelIsVisible: `${baseClassName}_label--visible`,
  container: `${baseClassName}_container`,
};

export interface CheckboxGroupProps {
  /** Custom classname to extend or customize */
  className?: string;
  /** Checkbox or ReactNode */
  children: Array<React.ReactElement<CheckboxProps> | ReactNode>;
  /** Hint text to be displayed under the label. */
  groupHintText?: string;
  /** Label for the group */
  labelText: string;
  /** Hide or show label. Label element is always present, but can be visually hidden.
   * @default visible
   */
  labelMode?: 'hidden' | 'visible';
  /** Text to mark a selection in group as optional. Will be wrapped in parentheses and shown after labelText. */
  optionalText?: string;
  /**
   * Unique id
   * If no id is specified, one will be generated
   */
  id?: string;
}

class BaseCheckboxGroup extends Component<
  CheckboxGroupProps & SuomifiThemeProp
> {
  render() {
    const {
      children,
      className,
      theme,
      labelText,
      labelMode,
      optionalText,
      groupHintText,
      id,
      ...passProps
    } = this.props;

    return (
      <HtmlDiv
        className={classnames(baseClassName, className)}
        id={id}
        {...passProps}
      >
        <HtmlFieldSet>
          <HtmlLegend className={checkboxGroupClassNames.legend}>
            <Label
              htmlFor={id}
              labelMode={labelMode}
              optionalText={optionalText}
              className={classnames({
                [checkboxGroupClassNames.labelIsVisible]:
                  labelMode !== 'hidden',
              })}
            >
              {labelText}
            </Label>
            {groupHintText && <HintText>{groupHintText}</HintText>}
          </HtmlLegend>
          <HtmlDiv className={checkboxGroupClassNames.container}>
            {children}
          </HtmlDiv>
        </HtmlFieldSet>
      </HtmlDiv>
    );
  }
}

const StyledCheckboxGroup = styled(BaseCheckboxGroup)`
  ${({ theme }) => baseStyles(theme)}
`;

/**
 * Use for grouping Checkboxes.
 */
export class CheckboxGroup extends Component<CheckboxGroupProps> {
  render() {
    const { id: propId, ...passProps } = this.props;
    return (
      <SuomifiThemeConsumer>
        {({ suomifiTheme }) => (
          <AutoId id={propId}>
            {(id) => (
              <StyledCheckboxGroup
                theme={suomifiTheme}
                id={id}
                {...passProps}
              />
            )}
          </AutoId>
        )}
      </SuomifiThemeConsumer>
    );
  }
}
