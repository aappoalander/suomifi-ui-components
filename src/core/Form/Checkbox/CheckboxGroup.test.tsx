import React from 'react';
import { render } from '@testing-library/react';
import { Checkbox } from './Checkbox';
import { CheckboxGroup } from './CheckboxGroup';

const CheckboxChilds = [1, 2, 3].map((value) => (
  <Checkbox key={value} id={`test-id-${value}`}>
    {`Label text ${value}`}
  </Checkbox>
));

describe('default, with only required props', () => {
  const DefaultGroup = (
    <CheckboxGroup labelText="Label">{CheckboxChilds}</CheckboxGroup>
  );

  it('should match snapshot', () => {
    const { container } = render(DefaultGroup);
    expect(container).toMatchSnapshot();
  });
});

describe('props', () => {
  describe('className', () => {
    const ClassNameGroup = (
      <CheckboxGroup labelText="Label" className="custom-className">
        {CheckboxChilds}
      </CheckboxGroup>
    );
    it('has the default className', () => {
      const { container } = render(ClassNameGroup);
      expect(container.firstChild).toHaveClass('fi-checkbox-group');
    });
    it('has the given className', () => {
      const { container } = render(ClassNameGroup);
      expect(container.firstChild).toHaveClass('custom-className');
    });
  });

  describe('groupHintText', () => {
    const HintTextGroup = (
      <CheckboxGroup labelText="Label" groupHintText="Example hint text">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has hintText element', () => {
      const { getByText } = render(HintTextGroup);
      const hintText = getByText('Example hint text');
      expect(hintText).toHaveClass('fi-hint-text');
    });
  });

  describe('id', () => {
    const IdGroup = (
      <CheckboxGroup labelText="Label" id="nice-id">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has the given id', () => {
      const { container } = render(IdGroup);
      expect(container.firstChild).toHaveAttribute('id', 'nice-id');
    });
  });

  describe('labelMode', () => {
    const LabelModeGroup = (
      <CheckboxGroup labelText="Label" labelMode="hidden">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has visually hidden label element', () => {
      const { getByText } = render(LabelModeGroup);
      const label = getByText('Label');
      expect(label).toHaveClass('fi-visually-hidden');
    });
  });

  describe('labelText', () => {
    const LabelTextGroup = (
      <CheckboxGroup labelText="Label">{CheckboxChilds}</CheckboxGroup>
    );

    it('has label element', () => {
      const { getByText } = render(LabelTextGroup);
      const label = getByText('Label');
      expect(label).not.toEqual(null);
    });
  });

  describe('optionalText', () => {
    const OptionalTextGroup = (
      <CheckboxGroup labelText="Label" optionalText="optional">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has optional text element', () => {
      const { getByText } = render(OptionalTextGroup);
      const optionalText = getByText('(optional)');
      expect(optionalText).not.toEqual(null);
    });
  });
});
