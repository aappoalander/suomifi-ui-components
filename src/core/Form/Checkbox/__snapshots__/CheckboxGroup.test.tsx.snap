// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default, with only required props should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c7 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c7::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c7::before,
.c7::after {
  box-sizing: border-box;
}

.c8 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c8::before,
.c8::after {
  box-sizing: border-box;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c5::before,
.c5::after {
  box-sizing: border-box;
}

.c9 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(0,0%,16%);
  font-size: 14px;
  line-height: 20px;
}

.c9.fi-status-text {
  display: block;
}

.c9.fi-status-text.fi-status-text--error {
  color: hsl(3,59%,48%);
}

.c6 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c6 .fi-checkbox_label {
  position: relative;
  display: block;
  padding-left: 25px;
  cursor: pointer;
  min-height: 27px;
  line-height: 1.5em;
  padding-top: 1px;
}

.c6 .fi-checkbox_label::before {
  content: '';
  position: absolute;
  left: 0px;
  top: 5px;
  box-sizing: border-box;
  height: 18px;
  width: 18px;
  border: 1px solid hsl(201,7%,58%);
  border-radius: 2px;
  background-color: hsl(0,0%,100%);
}

.c6 .fi-checkbox_icon {
  position: absolute;
  height: 10px;
  width: 10px;
  left: 4px;
  top: 9px;
}

.c6:focus-within .fi-checkbox_label::before {
  outline: 0;
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
}

.c6 .fi-checkbox_input {
  position: absolute;
  opacity: 0;
  z-index: -9999;
}

.c6 .fi-hint-text {
  padding-left: 25px;
  color: hsl(202,7%,40%);
}

.c6 .fi-status-text {
  line-height: 18px;
}

.c6 .fi-status-text.fi-checkbox_statusText--has-content {
  margin-top: 5px;
}

.c6.fi-checkbox--large .fi-checkbox_label {
  padding-left: 40px;
  min-height: 30px;
}

.c6.fi-checkbox--large .fi-checkbox_label::before {
  content: '';
  position: absolute;
  left: 0px;
  top: 0px;
  box-sizing: border-box;
  height: 30px;
  width: 30px;
  color: hsl(201,7%,58%);
  border: 2px solid;
}

.c6.fi-checkbox--large .fi-checkbox_label .fi-checkbox_icon {
  height: 20px;
  width: 20px;
  left: 5px;
  top: 4px;
}

.c6.fi-checkbox--large .fi-hint-text {
  padding-left: 40px;
}

.c6.fi-checkbox--checked .fi-checkbox_label::before {
  border-color: hsl(212,63%,45%);
}

.c6.fi-checkbox--checked .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(212,63%,45%);
}

.c6.fi-checkbox--error .fi-checkbox_label::before {
  border-color: hsl(3,59%,48%);
  border-width: 2px;
}

.c6.fi-checkbox--error .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(3,59%,48%);
}

.c6.fi-checkbox--disabled .fi-checkbox_label {
  cursor: not-allowed;
  color: hsl(202,7%,67%);
}

.c6.fi-checkbox--disabled .fi-checkbox_label::before {
  background-color: hsl(202,7%,97%);
  border-color: hsl(202,7%,80%);
  border-width: 1px;
}

.c6.fi-checkbox--disabled .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(202,7%,80%);
}

.c6.fi-checkbox--disabled.fi-checkbox--large .fi-checkbox_label::before {
  border-width: 2px;
}

.c4.fi-label-text .fi-label-text_label-span {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  display: block;
  color: hsl(0,0%,16%);
}

.c4.fi-label-text .fi-label-text_label-span .fi-label-text_optionalText {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c1 {
  color: hsl(0,0%,16%);
}

.c1 .fi-checkbox-group_label--visible {
  margin-bottom: 10px;
}

.c1 .fi-checkbox-group_legend .fi-hint-text {
  margin-bottom: 10px;
}

.c1 .fi-checkbox-group_container > .fi-checkbox {
  margin-bottom: 10px;
}

.c1 .fi-checkbox-group_container > .fi-checkbox:last-child {
  margin-bottom: 0;
}

.c1 .fi-checkbox-group_container .fi-checkbox--large {
  margin-top: 5px;
  margin-bottom: 15px;
}

<div>
  <div
    class="c0 fi-checkbox-group c1"
    id="1"
  >
    <fieldset
      class="c2"
    >
      <legend
        class="c3 fi-checkbox-group_legend"
      >
        <label
          class="c0 c4 fi-checkbox-group_label--visible fi-label-text"
          for="1"
        >
          <span
            class="c5 fi-label-text_label-span"
          >
            Label
          </span>
        </label>
      </legend>
      <div
        class="c0 fi-checkbox-group_container"
      >
        <div
          class="c0 fi-checkbox_container c6 fi-checkbox"
        >
          <input
            aria-invalid="false"
            class="c7 fi-checkbox_input"
            id="test-id-1"
            type="checkbox"
          />
          <label
            class="c8 fi-checkbox_label"
            for="test-id-1"
          >
            Label text 1
          </label>
          <span
            aria-atomic="true"
            aria-live="assertive"
            class="c5 c9 fi-status-text"
          />
        </div>
        <div
          class="c0 fi-checkbox_container c6 fi-checkbox"
        >
          <input
            aria-invalid="false"
            class="c7 fi-checkbox_input"
            id="test-id-2"
            type="checkbox"
          />
          <label
            class="c8 fi-checkbox_label"
            for="test-id-2"
          >
            Label text 2
          </label>
          <span
            aria-atomic="true"
            aria-live="assertive"
            class="c5 c9 fi-status-text"
          />
        </div>
        <div
          class="c0 fi-checkbox_container c6 fi-checkbox"
        >
          <input
            aria-invalid="false"
            class="c7 fi-checkbox_input"
            id="test-id-3"
            type="checkbox"
          />
          <label
            class="c8 fi-checkbox_label"
            for="test-id-3"
          >
            Label text 3
          </label>
          <span
            aria-atomic="true"
            aria-live="assertive"
            class="c5 c9 fi-status-text"
          />
        </div>
      </div>
    </fieldset>
  </div>
</div>
`;
